### PR TITLE
Python source

### PR DESCRIPTION
**Pull request by MrAnno**
## Python source

Two types of Python source can be implemented:

- Python `LogSource`:
  - You have 2 mandatory operations to implement, `run()` and `request_exit()`.
  - You can implement your own event loop, or integrate an external framework's or library's event loop (Kafka consumer, HTTP server, Flask, Twisted engine, etc).
  - You can post messages from `run()` using `LogSource::post_message()`.

- Python `LogFetcher`:
  - Easier to implement, the event loop and the entire scheduling is done automatically by syslog-ng.
  - You can use simple blocking server/client libraries to receive/fetch logs.
  - The following operations can be implemented: `open()`, `close()` and `fetch()`.

### Python `LogSource`
- A custom Python source implementation must be inherited from `syslogng.LogSource`. Multiple inheritance is allowed, but only for pure Python super classes.
- Mandatory methods: `run()`, `request_exit()`
- Optional methods: `init()`, `deinit()`, `suspend()`, `wakeup()`

`run()` can be used to implement an event loop or start a server framework/library. `LogMessage` instances can be created here. Messages can be sent using `LogSource::post_message()`.

`request_exit()` must shut down the event loop or framework, so the `run()` method can return gracefully. When blocking operations are used inside `run()`, `request_exit()` can be used to interrupt those operations and to set an exit flag.
`request_exit()` is called from a different thread (not from the source thread).

When `suspend()` and `wakeup()` are both implemented, they will be called when the source has to be suspended/resumed. If they are not specified, the entire source thread will be suspended automatically.

`suspend()` should prevent the source from posting new messages until `wakeup()` is called.
If this rule is violated, messages will be dropped with an error message: `Incorrectly suspended source, dropping message`.

Currently, `run()` stops permanently if an exception is propagated back to the C side (a `time-reopen()`-like retry mechanism can be implemented in the future).

```
    python(
        class("MySource")
        options(
            "option1" "value1",
            "option2" "value2"
        )
        # loaders("...")
        # log-iw-size(...)
        # persist-name(...)
        # flags(...)
    );
```

#### Example interface:
```python
from syslogng import LogSource
from syslogng import LogMessage
 
class MySource(LogSource):
    def init(self, options): # optional
        print("init")
        print(options)
        return True
 
    def deinit(self): # optional
        print("deinit")

    def run(self): # mandatory
        print("run")
        while not self.exit:
            # ...
            self.post_message(msg)
 
    def request_exit(self): # mandatory
        print("exit")
        self.exit = True
```

### Python `LogFetcher`

- A custom Python fetcher implementation must be inherited from `syslogng.LogFetcher`. 
- Mandatory methods: `fetch()`
- Optional methods: `init()`, `deinit()`, `open()`, `close()`, `request_exit()`
  - `request_exit()` is absolutely optional. You should use it only if `fetch()` has blocking operations and it has to be interrupted when shutting down syslog-ng. `request_exit()` is called from a different thread (not from the source thread).

image

(`connect()` means `open()` on the Python side, `disconnect()` = `close()`)

```
    python-fetcher(
        class("MyFetcher")
        options(
            "option1" "value1",
            "option2" "value2"
        )
        # loaders("...")
        # log-iw-size(...)
        # persist-name(...)
        # flags(...)
    );
```

#### Example interface
```python
from syslogng import LogFetcher
from syslogng import LogMessage
 
class MyFetcher(LogFetcher):
    def init(self, options): # optional
        print("init")
        print(options)
        return True
 
    def deinit(self): # optional
        print("deinit")
 
    def open(self): # optional
        print("open")
        return True
 
    def close(self): # optional
        print("close")
 
    def fetch(self): # mandatory
        print("fetch")
        # return LogFetcher.FETCH_ERROR,
        # return LogFetcher.FETCH_NOT_CONNECTED,
        return LogFetcher.FETCH_SUCCESS, msg
```

### `LogMessage` API

```python
msg = LogMessage() # empty message with good defaults (recvd timestamp, rcptid, hostid, ...)
msg = LogMessage("string or bytes-like object") # sets the $MESSAGE field

msg_ietf = LogMessage.parse('<165>1 2003-10-11T22:14:15.003Z mymachine.example.com evntslog - ID47 [exampleSDID@32473 iut="3" eventSource="Application" eventID="1011"] An application event log entry', self.parse_options)
msg_bsd = LogMessage.parse('<34>Oct 11 22:14:15 mymachine su: \'su root\' failed for lonvick on /dev/pts/8', self.parse_options)
 
msg["HOST"] = "hostname" # keep-hostname(), use-dns(), use-fqdn(), chain-hostnames() will be taken into account by the source
msg["MESSAGE"] = "message"
 
msg.set_pri(165)
 
timestamp = datetime.fromisoformat("2018-09-11T14:49:02.100+02:00")
msg.set_timestamp(timestamp) # datetime object, includes timezone information
```

Reusing/storing `LogMessage` objects after they have been posted (calling `post_message()` or returning the message from `fetch()`) is not recommended.

`set_timestamp()` works differently in Python 2 and Python 3.

- In Python 2, timezone information can not be attached to the datetime instance without using an external library. Naive datetime objects are supported, they are represented in UTC inside syslog-ng.
- In Python 3, naive and timezone-aware datetime objects are both supported.

These sources create structured `LogMessage` instances, so it is not necessarily a good idea to force `log-msg-size()` on the `MESSAGE` field or on the entire message. Currently, we can't return multiple messages, so splitting a message is not really possible, I would just ignore `log-msg-size()`.

----

Questions:

TODO:

- must haves:
  - [x] finish `LogMessage.parse("ietf or bsd raw message", self.parse_options)`
  - [x] finish `log_message.set_timestamp(ts)` (python2 compatibility is required)
  - [x] fix a IW-related crash
  - [x] `PyLogMessage` unit tests
- nice to haves
  - [ ] merge grammar keywords: `python()` and `python-fetcher()` --> `python()`
    - python-source and python-fetcher should be unified (1 source driver, 2 different worker implementations) so we can remove duplicated code and merge the grammar keywords as well.
  - [ ] implement `MainLoopWorker` timeout (shutdown)
  - [ ] `run()` time-reopen() restart when exception is propagated back to the C side

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/python/python-logmsg.c [hunk 1]
```diff
@@ -254,12 +256,63 @@ _logmessage_get_keys_method(PyLogMessage *self)
   return keys;
 }
 
+static PyLogMessage *
+py_log_message_set_pri(PyLogMessage *self, PyObject *args, PyObject *kwrds)
+{
```
Comment (Kokan) [inline modules/python/python-logmsg.c]: Do we want to return with the message in case of this functions ? I guess it would support chaining function calls, but I do not know if it is aligned with python own style ?!

File: modules/python/python-source.c [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+static gboolean
+_py_invoke_init(PythonSourceDriver *self)
+{
+  return _ps_py_invoke_bool_method_by_name_with_args(self, "init");
+}
+
+static void
+_py_invoke_deinit(PythonSourceDriver *self)
+{
+  _ps_py_invoke_void_method_by_name(self, "deinit");
```
Comment (Kokan) [inline modules/python/python-source.c]: I think it won't hurt to store `init`, `deinit`, ... as the `run_method`, because now this is going to print a debug method each time init/deinit is called, and not in startup. I would like to get a message about missing (even optional) methods when I start syslog-ng not when they happens to be called. (typo in function name hard to catch otherwise).

File: modules/python/python-fetcher.c [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+      return FALSE;
+    }
+}
+
+static LogThreadedFetchResult
+_py_invoke_fetch(PythonFetcherDriver *self)
+{
+  PyObject *ret = _py_invoke_function(self->py.fetch_method, NULL, self->class, self->super.super.super.super.id);
+
+  if (!ret || !PyTuple_Check(ret) || PyTuple_Size(ret) > 2)
```
Comment (Kokan) [inline modules/python/python-fetcher.c]: Why not assert for 2 item? Should we be okay without msg, even if some cases that should be always a dummy NULL?

File: modules/python/python-logmsg.c [hunk 4]
```diff
@@ -158,6 +158,29 @@ py_log_message_new(LogMessage *msg)
...
+  static const gchar *kwlist[] = {"message", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|z#", (gchar **) kwlist, &message, &message_length))
+    return NULL;
+
+  PyLogMessage *self = (PyLogMessage *) subtype->tp_alloc(subtype, 0);
+  if (!self)
+    return NULL;
+
+  self->msg = log_msg_new_empty();
+  invalidate_cached_time();
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/python/python-logmsg.c]: This was a point where I had enough :) The invalidation is here, because it is not called automatically, since it depends on ivykis. Threaded sources are not ivykis based, but I had to initialize ivykis, because all source threads require ivykis initialization because of the way we implemented the mark_freq timer...

TLDR: the invalidation is here because of mark messages :D  

We should definitely fix this whole thing.

File: modules/python/python-logmsg.c (hunk 1 above)
Comment (MrAnno) [reply modules/python/python-logmsg.c]: Yes, chaining was my intention. I'm not familiar with Python conventions, so I can remove this if it is unconventional. :)

File: modules/python/python-logmsg.c [hunk 5]
```diff
@@ -158,6 +158,29 @@ py_log_message_new(LogMessage *msg)
...
 
+static PyObject *
+py_log_message_new_empty(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
+{
+  const gchar *message = NULL;
+  gint message_length = 0;
+
+  static const gchar *kwlist[] = {"message", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|z#", (gchar **) kwlist, &message, &message_length))
+    return NULL;
```
[the remark this answers is not in the document]
Comment (Kokan) [reply modules/python/python-logmsg.c]: This was not understood by me. It is correct as it is, this sets the exception and returns with false.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/logthrsource/logthrsourcedrv.c [hunk 6]
```diff
@@ -121,19 +121,23 @@ void
 log_threaded_source_worker_options_defaults(LogThreadedSourceWorkerOptions *options)
 {
   log_source_options_defaults(&options->super);
+  msg_format_options_defaults(&options->parse_options);
+  options->parse_options.flags |= LP_SYSLOG_PROTOCOL;
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/logthrsource/logthrsourcedrv.c]: `LP_SYSLOG_PROTOCOL` can fall back to the legacy BSD format, so I think this is the good default.

File: modules/python/python-logmsg.c [hunk 7]
```diff
@@ -254,12 +254,25 @@ _logmessage_get_keys_method(PyLogMessage *self)
...
+{
+  guint pri;
+
+  static const gchar *kwlist[] = {"pri", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwrds, "I", (gchar **) kwlist, &pri))
+    return NULL;
+
+  self->msg->pri = pri;
+
+  Py_INCREF(self);
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/python/python-logmsg.c]: Return values have to be referenced.

For example, `Py_RETURN_NONE` is defined as
```
Py_INCREF(Py_None);
return Py_None;
```

File: modules/python/python-source.c [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+  if (!py_is_log_message((PyObject *) pymsg))
+    {
+      PyErr_Format(PyExc_TypeError, "LogMessage expected in the first parameter");
+      return NULL;
+    }
+
+  /* keep a reference until the PyLogMessage instance is freed */
+  LogMessage *message = log_msg_ref(pymsg->msg);
+  sd->post_message(sd, message);
+
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/python/python-source.c]: We shouldn't.

After pushing the message through the pipeline, it will be unrefed and freed by the destination.

The additional ref is required here because of the PyLogMessage instance.

When you create a new PyLogMessage, the LogMessage refcount will be set to 1.
This will be decreased to 0 in the destructor of PyLogMessage, if the message has not been sent out with post_message().

Before posting the message, an extra ref is required, because the corresponding PyLogMessage instance might be stored/cached/etc. on the Python side.

[the remark this answers is not in the document]
Comment (furiel) [reply modules/python/python-source.c]: Thanks for the explanation. And understand why we had the same ref in the fetcher case too.

File: modules/python/compat/compat-python-v3.c [hunk 9]
```diff
@@ -35,3 +38,48 @@ int_as_pyobject(gint num)
...
+    {
+      PyErr_Format(PyExc_TypeError, "datetime expected in the first parameter");
+      return FALSE;
+    }
+
+  PyObject *py_posix_timestamp = _py_invoke_method_by_name(py_timestamp, "timestamp", NULL,
+                                                           "PyDateTime", "py_datetime_to_logstamp");
+  if (!py_posix_timestamp)
+    {
+      PyErr_Format(PyExc_ValueError, "Error calculating POSIX timestamp");
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/python/compat/compat-python-v3.c]: `_py_invoke_method_by_name` does not "throw", it posts the exception to our internal log and clears it.
Throwing a new `PyExc_ValueError` exception is not the best idea, but I did not want to change the implementation of `_py_invoke_function`, it has too many users.

File: modules/python/python-source.c [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+  ((PyLogSource *) self->py.instance)->driver = self;
+
+  self->py.run_method = _py_get_attr_or_null(self->py.instance, "run");
+  self->py.request_exit_method = _py_get_attr_or_null(self->py.instance, "request_exit");
+  if (!self->py.run_method)
+    {
+      msg_error("Error initializing Python source, class does not have a run() method",
+                evt_tag_str("driver", self->super.super.super.id),
+                evt_tag_str("class", self->class));
+      return FALSE;
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/python/python-source.c]: Good catch, thanks!

The destination and the parser may have the same problem, I'll check them too.

**Comment by furiel** [comment]
After experimenting with this feature, let me share some thoughts.

LogFetcher:
- It is indeed a very convenient interface, It is nice avoid to implement an event loop, and the open/close methods are also convenient. However, a missing queue support limits the applicability somewhat. In some cases one fetch method would return many messages, typically if someone wants to implement a stream based source like file source or tcp source. Or some interfaces make messages available as bulks (like notifications in Github), or when for performance reason it would be good to fetch as many messages from the remote in one step as possible. In these cases, LogFetcher cannot be used, as it can only pass one message to the core. For these cases, the LogSource interface is more appropriate for now.
- I am not entirely sure how to halt the fetcher properly. Let's say I am in the middle of a long polling, when request exit arrives. Certainly, request exit can interrupt the long poll, but in the end I am left with no message to return. What should be the return value of the fetch? Is a `THREAD_FETCH_NOT_CONNECTED` is appropriate, or I need something else (concerned with statistics)? This would worth explaining this in the documentation and the pr description.

LogSource:
- The code looks very stable. I use this API for a loop(fetch-many-messages, sleep) use case.
- For a sleep based "eventloop", one would try to use `time.sleep`, but that is not very useful, because it cannot be interrupted. However I could get the sleep done with `event = threading.Event: event.wait(seconds)` will wait, but it can be interrupted with `event.set()`. See example below.
- Just reciting that is already written in the pr description, but the run method is somewhat fragile because the run method is one-shot. If for any reason the run method returns unexpectedly, either for a buggy eventloop implementation, or an uncatched exception, the whole source stops. At the same time, it is very to easy write a user code that is buggy and misses an exception. For example in case of a sleep-longpoll style source, or any fetch style source, the denial of service condition can go unnoticed for a long time. It would great if we could call run again if for some reason it exited (maybe together with a deinit-init pair). For now probably the best is to add a try-catch block inside the while True loop.

```
(import syslogng threading random)

(defclass PySource [syslogng.LogSource]
  (defn init [self options]
    (setv self.event (threading.Event)
          self.exit False)
    True)

  (defn run [self]
    (while (not self.exit)
      (when (random.randint 0 1)
        (raise (RuntimeError "its over")))
      (self.post_message (syslogng.LogMessage "example"))
      (self.event.wait 1)))

   (defn request_exit [self]
     (setv self.exit True)
     (self.event.set)))
```

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
@furiel Thanks for the review/tests.

> It is indeed a very convenient interface, It is nice avoid to implement an event loop, and the open/close methods are also convenient. However, a missing queue support limits the applicability somewhat. In some cases one fetch method would return many messages, typically if someone wants to implement a stream based source like file source or tcp source. Or some interfaces make messages available as bulks (like notifications in Github), or when for performance reason it would be good to fetch as many messages from the remote in one step as possible. In these cases, LogFetcher cannot be used, as it can only pass one message to the core. For these cases, the LogSource interface is more appropriate for now.

Yes, currently, you have to implement your own batching mechanism if you want to use LogFetcher. We could add some kind of batching support in the future.

> I am not entirely sure how to halt the fetcher properly. Let's say I am in the middle of a long polling, when request exit arrives. Certainly, request exit can interrupt the long poll, but in the end I am left with no message to return. What should be the return value of the fetch? Is a THREAD_FETCH_NOT_CONNECTED is appropriate, or I need something else (concerned with statistics)? This would worth explaining this in the documentation and the pr description.

You can return anything, or raise an exception. It's not cleanest thing, maybe we should add another non-error result to ThreadedFetcher and Python fetcher.

**Comment by MrAnno** [comment]
@kira-syslogng test this please test branch=pzolee-python-source;

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
@kira-syslogng test this please test branch=pzolee-python-source;

File: modules/python/python-fetcher.c [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+    case THREADED_FETCH_SUCCESS:
+      *result = (ThreadedFetchResult) ulong;
+      return TRUE;
+
+    default:
+      return FALSE;
+    }
+}
+
+static LogThreadedFetchResult
```
Comment (lbudai) [inline modules/python/python-fetcher.c]: What do you think about it: 
```C
static ThreadedFetchResult 
_fetch_log_message(PythonFetcherDriver *self, PyLogMessage **pymsg)
{
  ThreadedFetchResult fetch_result = THREADED_FETCH_ERROR; 
  *pymsg = NULL;

  PyObject *ret = _py_invoke_function(self->py.fetch_method, NULL, self->class, self->super.super.super.super.id);

  if (!ret || !PyTuple_Check(ret) || PyTuple_Size(ret) > 2)
    goto error;

  PyObject *result = PyTuple_GetItem(ret, 0); 
  if (!result || !PyLong_Check(result))
    goto error;

  if (!_ulong_to_fetch_result(PyLong_AsUnsignedLong(result), &fetch_result))
    goto error;

  *pymsg = (PyLogMessage *) PyTuple_GetItem(ret, 1);
   if (!*pymsg || !py_is_log_message((PyObject *) pymsg))
     goto error;

   goto exit_;
error:
  msg_error("Error in Python fetcher, fetch() must return a tuple (FetchResult, LogMessage)",
            evt_tag_str("driver", self->super.super.super.super.id),
            evt_tag_str("class", self->class));

exit_:
  Py_XDECREF(ret);
  PyErr_Clear();
  return fetch_result;
}

static LogThreadedFetchResult
_py_invoke_fetch(PythonFetcherDriver *self)
{
  PyLogMessage *pymsg = NULL;
 
  LogThreadedFetchResult fetch_result = { .msg = NULL };
  fetch_result.result = _fetch_log_message(self, &pymsg);

  if (fetch_result.result == THREADED_FETCH_SUCCESS)
    {
      /* keep a reference until the PyLogMessage instance is freed */
      fetch_result.msg = log_msg_ref(pymsg->msg);
    }

 return fetch_result;
}
```
?

File: modules/python/python-fetcher.c [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+static gboolean
+_py_is_log_fetcher(PyObject *obj)
+{
+  return PyType_IsSubtype(Py_TYPE(obj), &py_log_fetcher_type);
+}
+
+static gboolean
+_py_init_bindings(PythonFetcherDriver *self)
```
Comment (lbudai) [inline modules/python/python-fetcher.c]: split into smaller functions

File: modules/python/python-fetcher.c [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+{
+  PythonFetcherDriver *self = (PythonFetcherDriver *) s;
+
+  PyGILState_STATE gstate = PyGILState_Ensure();
+  _py_invoke_request_exit(self);
+  PyGILState_Release(gstate);
+}
+
+static gboolean
+python_fetcher_init(LogPipe *s)
```
Comment (lbudai) [inline modules/python/python-fetcher.c]: split into smaller functions.

**Comment by lbudai** [comment]
Some refactors needed, but I like this work :) 
And the description part of the PR is really nice, thanks!

**Comment by MrAnno** [comment]
@kira-syslogng test this please test branch=pzolee-python-source;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
@kira-syslogng test this please test branch=pzolee-python-source;

**Comment by MrAnno** [comment]
@kira-syslogng do perftest

**Comment by MrAnno** [comment]
@kira-syslogng test this please test branch=pzolee-python-source;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
@kira-syslogng do perftest

**Comment by kira-syslogng** [comment]
Kira-performance-test: Build SUCCESS
